### ISSUE TITLE
Customer Segmentation & Prediction

### ISSUE THREAD

**Issue by amrutadpatil2001-droid** [amrutadpatil2001-droid/Python_Projects#14]
import pandas as pd
import numpy as np

# Load dataset
df = pd.read_csv("customer_churn.csv")

# Basic check
print(df.head())
print(df.info())
print(df.describe())

# Handle missing values
df = df.dropna()
# Convert categorical → numerical
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
for col in df.select_dtypes(include='object').columns:
    df[col] = le.fit_transform(df[col])

# Feature scaling
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(df)

from sklearn.cluster import KMeans
import matplotlib.pyplot as plt
wcss = []

for i in range(1, 11):
    kmeans = KMeans(n_clusters=i, random_state=42)
    kmeans.fit(X_scaled)
    wcss.append(kmeans.inertia_)
plt.plot(range(1,11), wcss)
plt.xlabel("Clusters")
plt.ylabel("WCSS")
plt.title("Elbow Method")
plt.show()

kmeans = KMeans(n_clusters=3, random_state=42)
df['Cluster_KMeans'] = kmeans.fit_predict(X_scaled)

from scipy.cluster.hierarchy import dendrogram, linkage
linked = linkage(X_scaled, method='ward')
dendrogram(linked)
plt.show()

from sklearn.cluster import AgglomerativeClustering
hc = AgglomerativeClustering(n_clusters=3)
df['Cluster_HC'] = hc.fit_predict(X_scaled)

from scipy.cluster.hierarchy import dendrogram, linkage
linked = linkage(X_scaled, method='ward')
dendrogram(linked)
plt.show()

from sklearn.cluster import AgglomerativeClustering
hc = AgglomerativeClustering(n_clusters=3)
df['Cluster_HC'] = hc.fit_predict(X_scaled)

from sklearn.cluster import DBSCAN
dbscan = DBSCAN(eps=0.5, min_samples=5)
df['Cluster_DBSCAN'] = dbscan.fit_predict(X_scaled)
cluster_summary = df.groupby('Cluster_KMeans').mean()
print(cluster_summary)

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
models = {}
for cluster in df['Cluster_KMeans'].unique():
    segment = df[df['Cluster_KMeans'] == cluster]
    X = segment.drop(['Churn','Cluster_KMeans'], axis=1)
    y = segment['Churn']
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
    model = RandomForestClassifier(random_state=42)
    model.fit(X_train, y_train)
    models[cluster] = (model, X_test, y_test)

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
for cluster, (model, X_test, y_test) in models.items():
    y_pred = model.predict(X_test)
    print(f"\nCluster {cluster}")
    print("Accuracy:", accuracy_score(y_test, y_pred))
    print("Precision:", precision_score(y_test, y_pred))
    print("Recall:", recall_score(y_test, y_pred))
    print("F1 Score:", f1_score(y_test, y_pred))

from sklearn.model_selection import GridSearchCV
param_grid = {
    'n_estimators': [50, 100],
    'max_depth': [5, 10, None],
    'min_samples_split': [2, 5]
}
grid = GridSearchCV(RandomForestClassifier(), param_grid, cv=3)
grid.fit(X_train, y_train)
print("Best Params:", grid.best_params_)

importances = model.feature_importances_
features = X.columns
importance_df = pd.DataFrame({
    'Feature': features,
    'Importance': importances
}).sort_values(by='Importance', ascending=False)
print(importance_df.head(10))
